cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/services/SourceControlManagementService.java

```diff
@@ -285,4 +288,18 @@ public RepositoryAppsResponse listApps(NamespaceId namespace) throws NotFoundExc
     RepositoryConfig repoConfig = getRepositoryMeta(namespace).getConfig();
     return sourceControlOperationRunner.list(new NamespaceRepository(namespace, repoConfig));
   }
+
+  @Override
+  protected void startUp() throws Exception {
+    if (sourceControlOperationRunner instanceof AbstractIdleService) {
```

Comment:
instanceof should normally be avoided if possible. It is easy to break with future refactoring. We generally also want to avoid starting/stopping services in other services, as it can get a bit strange if multiple services all use each other.

In this particular case, I think we can make SourceControlOperationRunner extend Service, where the remote one does nothing and in-memory sets the SystemReader. Bind it to a singleton, and start it in AppFabricServer instead of SourceControlManagementService (which no longer needs to extend AbstractIdleService).

Reply:
Followed the approach of extending the `Service` interface in `SourceControlOperationRunner`.